### PR TITLE
network: add RoutingPolicyRule.Family= setting

### LINKED ISSUE #13233: RoutingPolicyRule has no knob to create AF_INET6 rule
**systemd version the issue has been seen with**
> 242

**Used distribution**
 > NixOS

**Expected behaviour you didn't see**
 > networkd should be able to add AF_INET6 policy rules (`ip -6 rule`)

**Unexpected behaviour you saw**
 > networkd is hardcoded to create rules for AF_INET only, while most of the code seems to handle AF_INET6.

https://github.com/systemd/systemd/blob/4ff296b02411bb4f0dc38f48cbab06f8645d2a08/src/network/networkd-routing-policy-rule.c#L25-L28

**Steps to reproduce the problem**

```
[RoutingPolicyRule]
IncomingInterface=vlan200
Table=100
Priority=100
```

will always result in an AF_INET rule, there is no knob to make it apply to AF_INET6 traffic.

```
# ip ru
0:	from all lookup local
100:	from all iif vlan200 lookup 100
32766:	from all lookup main
32767:	from all lookup default
# ip -6 ru
0:	from all lookup local
32766:	from all lookup main
```

### PR DESCRIPTION
**Pull request by yuwata**
Closes #13233.

cc @ssahani.

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -1139,6 +1139,15 @@
             <para>A boolean. Specifies whether the rule to be inverted. Defaults to false.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>Family=</varname></term>
+          <listitem>
+            <para>Takes a special value <literal>AF_INET</literal> or <literal>AF_INET6</literal>.
```
Comment (ssahani) [inline man/systemd.network.xml]: ```AF_INET``` is like programmatically. We should rename this to ```ipv4``` /  ```ipv6```

**Comment by mweinelt** [comment]
Tested and works, thank you.

```
[RoutingPolicyRule]
IncomingInterface=vlan200
Table=100
Priority=100
Family=AF_INET6
```

```
# ip -6 ru
0:	from all lookup local
100:	from all iif vlan200 lookup 100
32766:	from all lookup main
```

**Comment by yuwata** [comment]
@ssahani Thank you for the comment. I've changed `AF_INET` and `AF_INET6` to `ipv4` and `ipv6`, respectively.

@mweinelt Thank you for testing this PR!

**Comment by flo-wer** [comment]
Maybe add a `both` or `any` value to use the rule for IPv4 and IPv6? This might be a common use case if the address is omitted.

File: man/systemd.network.xml [hunk 2]
```diff
@@ -1139,6 +1139,15 @@
             <para>A boolean. Specifies whether the rule to be inverted. Defaults to false.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>Family=</varname></term>
+          <listitem>
+            <para>Takes a special value <literal>ipv4</literal> or <literal>ipv6</literal>.
```
Comment (keszybz) [inline man/systemd.network.xml]: Yeah, it'd be nice to change this to: "Takes either "ipv4", "ipv6", or "both"." (and of course adjust the code to match that).

**Comment by keszybz** [comment]
I think we can merge this even for v243, since it's simple enough and fixes an obvious annoyance.

**Comment by yuwata** [comment]
@flo-wer and @keszybz Thank you for the comments. Now 'both' is also supported. PTAL.

**Comment by ssahani** [comment]
I think both means we need to apply for ```ipv4``` and ```ipv6``` . That means the rules needed to be added twice  @keszybz  ?

**Comment by mrc0mmand** [comment]
LSan is not happy:

```
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]: ==6854==ERROR: LeakSanitizer: detected memory leaks
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]: Direct leak of 272 byte(s) in 2 object(s) allocated from:
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #0 0x7f9963a1835a in malloc (/usr/lib/clang/8.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x13035a)
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #1 0x560a296a7eae in routing_policy_rule_new /build/build/../src/network/networkd-routing-policy-rule.c:22:16
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #2 0x560a296b58da in routing_policy_load_rules /build/build/../src/network/networkd-routing-policy-rule.c:1186:21
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #3 0x560a2967a650 in manager_new /build/build/../src/network/networkd-manager.c:1379:16
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #4 0x560a29670385 in run /build/build/../src/network/networkd.c:74:13
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #5 0x560a2966fec3 in main /build/build/../src/network/networkd.c:121:1
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #6 0x7f99629f6ee2 in __libc_start_main (/lib64/libc.so.6+0x26ee2)
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]: Indirect leak of 12 byte(s) in 2 object(s) allocated from:
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #0 0x7f99639a90ae in __interceptor_strdup (/usr/lib/clang/8.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xc10ae)
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #1 0x7f9963201592 in free_and_strdup /build/build/../src/basic/string-util.c:999:21
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #2 0x560a296b59c7 in routing_policy_load_rules /build/build/../src/network/networkd-routing-policy-rule.c:1242:37
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #3 0x560a2967a650 in manager_new /build/build/../src/network/networkd-manager.c:1379:16
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #4 0x560a29670385 in run /build/build/../src/network/networkd.c:74:13
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #5 0x560a2966fec3 in main /build/build/../src/network/networkd.c:121:1
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]:     #6 0x7f99629f6ee2 in __libc_start_main (/lib64/libc.so.6+0x26ee2)
Aug 01 15:54:05 arch.localdomain systemd-networkd[6854]: SUMMARY: AddressSanitizer: 284 byte(s) leaked in 4 allocation(s).


Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]: ==6884==ERROR: LeakSanitizer: detected memory leaks
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]: Direct leak of 272 byte(s) in 2 object(s) allocated from:
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #0 0x7f2ec99c835a in malloc (/usr/lib/clang/8.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x13035a)
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #1 0x5584c920ceae in routing_policy_rule_new /build/build/../src/network/networkd-routing-policy-rule.c:22:16
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #2 0x5584c921a8da in routing_policy_load_rules /build/build/../src/network/networkd-routing-policy-rule.c:1186:21
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #3 0x5584c91df650 in manager_new /build/build/../src/network/networkd-manager.c:1379:16
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #4 0x5584c91d5385 in run /build/build/../src/network/networkd.c:74:13
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #5 0x5584c91d4ec3 in main /build/build/../src/network/networkd.c:121:1
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #6 0x7f2ec89a6ee2 in __libc_start_main (/lib64/libc.so.6+0x26ee2)
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]: Indirect leak of 12 byte(s) in 2 object(s) allocated from:
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #0 0x7f2ec99590ae in __interceptor_strdup (/usr/lib/clang/8.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xc10ae)
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #1 0x7f2ec91b1592 in free_and_strdup /build/build/../src/basic/string-util.c:999:21
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #2 0x5584c921a9c7 in routing_policy_load_rules /build/build/../src/network/networkd-routing-policy-rule.c:1242:37
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #3 0x5584c91df650 in manager_new /build/build/../src/network/networkd-manager.c:1379:16
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #4 0x5584c91d5385 in run /build/build/../src/network/networkd.c:74:13
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #5 0x5584c91d4ec3 in main /build/build/../src/network/networkd.c:121:1
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]:     #6 0x7f2ec89a6ee2 in __libc_start_main (/lib64/libc.so.6+0x26ee2)
Aug 01 15:54:19 arch.localdomain systemd-networkd[6884]: SUMMARY: AddressSanitizer: 284 byte(s) leaked in 4 allocation(s).


 ____________________________________________
/ Found   2 sanitizer errors (  2 ASan,   0  \
| UBSan,   0 MSan). Looks like you need to   |
\ look at the log                            /
 --------------------------------------------
 \
  \
     __
    /  \
    |  |
    @  @
    |  |
    || |/
    || ||
    |\_/|
    \___/
```

**Comment by yuwata** [comment]
> I think both means we need to apply for `ipv4` and `ipv6` . That means the rules needed to be added twice @keszybz ?

@ssahani Yes. Please see `routing_policy_rule_section_verify()`.

> LSan is not happy:

@mrc0mmand Thanks. I've added fixes for that. I hope now CIs are all green.

File: src/network/networkd-util.c [hunk 3]
```diff
@@ -24,8 +24,15 @@ static const char * const link_local_address_family_boolean_table[_ADDRESS_FAMIL
         [ADDRESS_FAMILY_FALLBACK_IPV4] = "ipv4-fallback",
 };
 
+static const char * const routing_policy_rule_address_family_boolean_table[_ADDRESS_FAMILY_BOOLEAN_MAX] = {
```
Comment (keszybz) [inline src/network/networkd-util.c]: Why is this called boolean? That seems rather misleading. I see it is for consistency with surrounding code, but I think this should be changed (everywhere) at some point.

File: src/network/networkd-routing-policy-rule.c [hunk 4]
```diff
@@ -552,12 +553,60 @@ int routing_policy_rule_configure(RoutingPolicyRule *rule, Link *link, link_netl
...
+                        return r;
+
+                n->invert_rule = rule->invert_rule;
+                n->tos = rule->tos;
+                n->protocol = rule->protocol;
+                n->table = rule->table;
+                n->fwmark = rule->fwmark;
+                n->fwmask = rule->fwmask;
+                n->priority = rule->priority;
+                n->family = AF_INET6;
```
Comment (keszybz) [inline src/network/networkd-routing-policy-rule.c]: Should this be always `AF_INET6`?

Comment (yuwata) [reply src/network/networkd-routing-policy-rule.c]: Yes. When `Family=both`, we need to copy the section. The original one becomes `AF_INET`, so the new one must be `AF_INET6`.

File: src/network/networkd-util.c (hunk 3 above)
Comment (yuwata) [reply src/network/networkd-util.c]: OK. I will rename them.

**Comment by yuwata** [comment]
@keszybz Updated.
- `routing_policy_rule_copy()` is introduced. Then the copying logic in `routing_policy_rule_section_verify()` becomes easy to understand.
- `AddressFamilyBoolean` -> `AddressFamily`

PTAL.

**Comment by yuwata** [comment]
Rebased. PTAL.

**Comment by keszybz** [comment]
Looks great, thanks.

**Pull request merged**